specs/network/cases.md

```diff
@@ -0,0 +1,128 @@
+// know the list of prefixes that need to be routed to that IPv6 address.
+// An upstream router is the entry point toward nodes that have only IPv6 acces
+// throug tunnels (like nodes in ipv4-only networks or home networks)
+type exit {
+    // netresource is the same as on all other netresources of a tenant network
+    netresource
+    // the ultimate IPv{4,6} config of the exit container.
+    ipv4_conf
+    ipv6_conf
+}
+
+type ipv4_conf {
+    // cidr
+    ipaddress 
+    gateway
+    metric
+    // deterministic name in function of the prefix and it's allocation
+    interface
+    // TBD, we need to establish if we want fc00/7 (ULA) or rfc1918 networks 
+    // to be NATed (6t04 and/or 66)
+    nat
+}
+// comments: see above
+type ipv6_conf {
+    ipaddress // cidr
+    gateway
+    metric
+    interface
+}
+
+// connected are the peers for which we have a tunnel establised, and the
+// prefix it routes to. The connection, as it is a peer to peer connection,
+// can be of type wireguard, but also any other type that can bring
+// a packet to a node containing a netresource.
+// If for instance that node lives in the same subnet, it'll be a lot more
+// efficient to set up a vxlan (multicast or with direct fdb entries), than
+// using wireguard tunnels (that can be seen in a later phase)
+type connected {
+    connection
+    prefix
+}
+
+// see above
+enum connection {
+    wireguard
+    localvxlan
+}
+
+// see above for the type definition.
+// the key would be a public key, with the private key only available
+// locally and stored locally. 
+type wireguard {
+    // deterministic, based on the public key
+    nicname
+    // TBD, a peer can be IPv6, IPv6-ll or IPv4
+    peer
+    // public key
+    key
+    // Gateways in IPv6 are link-local. To be able to use IPv6 in any way,
+    // an interface needs an IPv6 link-local address. As wireguard interfaces 
+    // are l3-only, the kernel doesn't assign one, so we need to assign one
+    // ourselves. (we need to come up with a deterministic way, so we can be
+    // sure we now which/where)
+    ll_fe80
+}
+
+// definition for later usage
+// an l2vxlan wil be connected to a default bridge that gets attached to the
+// network resource. That way we can easily add a vxlan to that bridge for 
+// local interconnectivity
+type l2vxlan {
+    // deterministic or stored... 
+    nicname
+    // Or it's through fdb entries 
+    Option<peer>
+    // Or it's in a multicast vxlan
+    Option<group>
+    // a vxlan always has an ID
+    id
+}
+
+```
+
+### QUESTIONS:
+
+  - Storage of network objects.
+  It has been said that we will use the 'blockchain' to store 'things'. A network/network resource could very well be such a 'thing', but at the moment there is no clear view on how to do that and how big these 'things' can be. Also, that 'thing' needs to be encrypted to make sure we don't enlarge a possible attack surface towards a tenant network.
```

Comment:
As far as I understand the "blockchain" used to store data won't be a "real" blockchain, but just a bunch of BCDB with some RAFT on top for consensus.
So storage wise we should not have a problem cause BCDB can store big amount of data.
Now for notification, seems we need to think how we could create something that doesn't require all the node too keep doing some pooling on the BCDB, that would requires way to much bandwidth.
Maybe we could open a FR in BCDB to have some kind of publish/subcribe mechanism so any change can be pushed the the required nodes.
Needs further thinking on this.